Add status icons

freakboy3742 opened the pull request:
Adds an API allowing users to register status icons with an app.

* Adds a `status_icons` attribute to app. Status icons can be registered with the app, and can provided with an icon (falling back to the app's icon if no icon is provided), and a text label (falling back to the app name).
* Adds a `StatusIcon` class, representing a "button-like" status icon. This has an `on_press` handler.
* Adds a `MenuStatusIcon` class, representing a "menu-like" status icon. 
* A `MenuStatusIcon` can also act as a group, so groups can be created using the MenuStatusIcon as a parent. 
* The status icons list has a `commands` CommandSet. Commands that you want to appear in status menus are registered with that CommandSet. 
* Any command registered against the `COMMANDS` group will be added to the first MenuStatusIcon that is added to  `status_icons`. This allows for a safe default if you define commands without a specific status icon group.
* Standard commands (currently About and Exit) are added to `status_icons.commands`, and registered as part of the `COMMANDS` group, so they will be added to the first MenuStatusIcon for the app.
* Any command that is part of a group that *isn't* added to `status_icons`, and isn't `COMMANDS`, will be ignored.

There's one TODO in the implementation at present, waiting on the new "Standard command" API from #2666.

There is a stub implementation on Android, iOS, Web and Textual. This allows apps that *define* status icons to at least *start*, albeit with a warning. 

The GTK implementation uses XApp, rather than AyatanaAppIndicator3 as discussed in the ticket. XApp uses almost the same API as Ayatana, but is being developed by Mint who (unlike Ubuntu) actually *do* believe in status icons, and as a result, the implementation actually works as you'd expect.

Fixes #97 

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented on textual/src/toga_textual/statusicons.py:
```diff
@@ -0,0 +1,32 @@
+        pass
+
+    def create(self):
+        toga.NotImplementedWarning.warn("Android", "Status Icons")
```
```suggestion
        toga.NotImplementedWarning.warn("Textual", "Status Icons")
```

mhsmith commented:
> * Any command that is part of a group that _isn't_ added to `status_icons`, and isn't `COMMANDS`, will be ignored.

Since this must be a programming error, rather than failing silently I think it would make sense to throw an exception so the developer is pointed directly at the cause.

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+A simple status icon is a bare icon in the status bar. You can set and change the icon
+as required to reflect changes in application state; by default, the status icon will
+use the app's icon. The text associated with the icon will be used as a tooltip; again,
+the apps formal name will be used as default text. The icon can respond to mouse clicks
```
```suggestion
the app's formal name will be used as default text. The icon can respond to mouse clicks
```

mhsmith commented on core/src/toga/app.py:
```diff
@@ -653,6 +664,11 @@ def commands(self) -> CommandSet:
         return self._commands
 
+    @property
+    def status_icons(self) -> CommandSet:
```
```suggestion
    def status_icons(self) -> StatusIconSet:
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+by defining an ``on_press`` handler.
+
+To define a simple status icon, declare an instance of :class:`toga.StatusIcon`, and add
+it to your app's :attr:`toga.App.status_icons` set:
```
```suggestion
it to your app's :attr:`~toga.App.status_icons` set:
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+The :class:`~toga.MenuStatusIcon` is a :class:`~toga.Group` for command sorting
+purposes. To put a command in a menu associated with a :class:`~toga.MenuStatusIcon`,
+set the ``group`` associated with that command, then add the command to the
+:class:`~toga.commands.CommandSet` associated with status icons. The following example
```
```suggestion
:class:`~toga.command.CommandSet` associated with status icons. The following example
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+    # Create a MenuStatusIcon
+    status_icon = MenuStatusIcon(icon="icons/blue")
+
+    # Create some commands that are associated with the # menu status icon's group.
```
```suggestion
    # Create some commands that are associated with the menu status icon's group.
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+
+If you add at least one :class:`~toga.MenuStatusIcon` instance to your app, Toga will
+add some standard commands to the app's status icon command set. These items will appear
+at the end of menu of the first :class:`~toga.MenuStatusIcon`'s menu. To remove these
```
```suggestion
at the end of the first :class:`~toga.MenuStatusIcon`'s menu. To remove these
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+Reference
+---------
+
+.. autoclass:: toga.statusicons.BaseStatusIcon
+
+.. autoclass:: toga.StatusIcon
+
+.. autoclass:: toga.MenuStatusIcon
```
It's a little confusing that a `MenuStatusIcon` is not a `StatusIcon`. I think it would be clearer if the base class was called `StatusIcon` and the simple class was called`SimpleStatusIcon`.

freakboy3742 replied:
Agreed - I had the same reservation about naming.

mhsmith commented on core/src/toga/statusicons.py:
```diff
@@ -0,0 +1,247 @@
+    def __repr__(self):
+        return f"<StatusIcon {self.text!r}: {self.id}>"
+
+    @property
+    def id(self) -> str:
+        """A unique identifier for the icon."""
+        return self._id
+
+    @property
+    def text(self) -> str:
+        """A text label for the icon."""
+        return self._text
```
I understand it's been done this way because of the duplicate property names in `Group`, but for the purposes of the documentation, the `id` and `text` properties should appear in the base class.

mhsmith commented on core/src/toga/statusicons.py:
```diff
@@ -0,0 +1,247 @@
+class BaseStatusIcon:
+    def __init__(self, icon: IconContentT | None = None, **kwargs):
+        """An abstract base class for all status icons."""
+        super().__init__(**kwargs)
```
In isolation it's not obvious why a top-level class would call `super`. It would be clearer for `MenuStatusIcon` to call `__init__` on both of its base classes directly, without using `super`. Then these `**kwargs` can be removed.

freakboy3742 replied:
Yeah - MRO is a bit of a pain with complex hierarchies. I went with the "least specific" implementation, but I agree that explicitly calling the constructors you want is more obvious on first reading.

mhsmith commented on core/src/toga/statusicons.py:
```diff
@@ -0,0 +1,247 @@
+        self.elements: dict[str, BaseStatusIcon] = {}
+        self.commands = CommandSet()
+
+    @property
+    def menu_status_icons(self):
+        """An iterator over the menu status icons that have been registered."""
+        return (icon for icon in self if isinstance(icon, MenuStatusIcon))
+
+    @property
+    def primary_menu_status_icon(self):
+        """The first menu status icon that has been registered.
```
Do these need to be public? I can't think of why an app would use them.

freakboy3742 replied:
I can't think of a particular use case for them; but I couldn't think of a reason to *not* include them either. However, you're probably right that it's safer to hide them initially and make them public when/if there's a use case.

mhsmith commented on core/src/toga/statusicons.py:
```diff
@@ -0,0 +1,247 @@
+
+        :param status_icon: The status icon instance to remove.
+        :raises ValueError: If the status icon commands include any commands that
+            reference the icon that has been removed, a ValueError error will be raised.
```
```suggestion
            reference the icon that has been removed.
```

mhsmith commented on core/src/toga/statusicons.py:
```diff
@@ -0,0 +1,247 @@
+    def clear(self):
+        """Remove all the icons from the set.
+
+        :raises ValueError: If the status icon commands include any commands that
+            reference the icon that has been removed, a ValueError error will be raised.
```
```suggestion
        :raises ValueError: If the status icon commands include any commands that
            reference an icon that has been removed.
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -61,23 +61,23 @@ use the app's icon. The text associated with the icon will be used as a tooltip;
     # Define a status icon that uses default values for icon and tooltip,
     # and doesn't respond to mouse clicks.
-    status_icon_1 = StatusIcon()
+    status_icon_1 = SimpleStatusIcon()
```
```suggestion
    status_icon_1 = toga.SimpleStatusIcon()
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -61,23 +61,23 @@ use the app's icon. The text associated with the icon will be used as a tooltip;
         print("Second status icon pressed!")
 
-    status_icon_2 = StatusIcon(
+    status_icon_2 = SimpleStatusIcon(
```
```suggestion
    status_icon_2 = toga.SimpleStatusIcon(
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+Menu status icons
+~~~~~~~~~~~~~~~~~
+
+A menu-based status icon is define by adding a :class:`toga.MenuStatusIcon` instance. A
```
```suggestion
A menu-based status icon is defined by adding a :class:`toga.MenuStatusIcon` instance. A
```

mhsmith commented on docs/reference/api/resources/statusicons.rst:
```diff
@@ -0,0 +1,187 @@
+.. code-block:: python
+
+    # Create a MenuStatusIcon
+    status_icon = MenuStatusIcon(icon="icons/blue")
```
```suggestion
    status_icon = toga.MenuStatusIcon(icon="icons/blue")
```

freakboy3742 commented:
I've incorporated those doc updates, merged with main, replaced the TODO about standard commands with the actual standard command mechanism, and updated the tests and screenshots to reflect the standard menu items. I think that makes this ready for final review and merge.
